Document/0x06h-Testing-Platform-Interaction.md

```diff
@@ -3,365 +3,2753 @@
+    log("\t\tuserInfo: " + ObjC.Object(args[3]).userInfo().toString());
+    log("\trestorationHandler: " +ObjC.Object(args[4]).toString());
+  },
+```
+
+The new output is:
+
+```
+298382 ms  -[AppDelegate application:0x10556b3c0 continueUserActivity:0x1c4237780 
+                restorationHandler:0x16f27a898]
+298382 ms  	application:<Application: 0x10556b3c0>
+298382 ms  	continueUserActivity:<NSUserActivity: 0x1c4237780>
+298382 ms  		webpageURL:http://t.me/addstickers/radare
+298382 ms  		activityType:NSUserActivityTypeBrowsingWeb
+298382 ms  		userInfo:{
+}
+298382 ms  	restorationHandler:<__NSStackBlock__: 0x16f27a898>
+```
+
+Apart from the function parameters we have added more information by calling some methods from them to get more details, in this case about the `NSUserActivity`. If we look in the [Apple Developer Documentation](https://developer.apple.com/documentation/foundation/nsuseractivity?language=objc "NSUserActivity") we can see what else we can call from this object.
+
+###### Checking How the Links Are Opened
+
+If you want to know more about which function actually opens the URL and how the data is actually being handled you should keep investigating.
+
+Extend the previous command in order to find out if there are any other functions involved into opening the URL.
+
+```bash
+$ frida-trace -U Telegram -m "*[* *restorationHandler*]" -i "*open*Url*"
+```
+
+> `-i` includes any method. You can also use a glob pattern here (e.g. `-i "*open*Url*"` means "include any function containing 'open', then 'Url' and something else")
+
+Again, we let first frida-trace generate the stubs in `__handlers__/`:
+
+```
+$ frida-trace -U Telegram -m "*[* *restorationHandler*]" -i "*open*Url*"
+Instrumenting functions...
+-[AppDelegate application:continueUserActivity:restorationHandler:]
+$S10TelegramUI0A19ApplicationBindingsC16openUniversalUrlyySS_AA0ac4OpenG10Completion...
+$S10TelegramUI15openExternalUrl7account7context3url05forceD016presentationData18application...
+$S10TelegramUI31AuthorizationSequenceControllerC7account7strings7openUrl5apiId0J4HashAC0A4Core19...
+...
+```
+
+Now you can see a long list of functions but we still don't know which ones will be called. Trigger the universal link again and observe the traces.
+
+```
+           /* TID 0x303 */
+298382 ms  -[AppDelegate application:0x10556b3c0 continueUserActivity:0x1c4237780 
+                restorationHandler:0x16f27a898]
+298619 ms     | $S10TelegramUI15openExternalUrl7account7context3url05forceD016presentationData
+                18applicationContext20navigationController12dismissInputy0A4Core7AccountC_AA
+                14OpenURLContextOSSSbAA012PresentationK0CAA0a11ApplicationM0C7Display0
+                10NavigationO0CSgyyctF()
+```
+
+Apart from the Objective-C method, now there is one Swift function that is also of your interest.
+
+There is probably no documentation for that Swift function but you can just demangle its symbol using `swift-demangle`:
```

Comment:
I don't think xcrun has been mentioned before, and I hadn't hard about swift-demangle yet myself. Maybe we should add some info, or at least a reference to available commands?

Reply:
done